volatility3/framework/plugins/windows/devicetree.py

```diff
@@ -0,0 +1,145 @@
+    0x0000001a : "FILE_DEVICE_SERIAL_MOUSE_PORT",
+    0x0000001b : "FILE_DEVICE_SERIAL_PORT",
+    0x00000031 : "FILE_DEVICE_SMARTCARD",
+    0x0000002e : "FILE_DEVICE_SMB",
+    0x0000001d : "FILE_DEVICE_SOUND",
+    0x0000001e : "FILE_DEVICE_STREAMS",
+    0x0000001f : "FILE_DEVICE_TAPE",
+    0x00000020 : "FILE_DEVICE_TAPE_FILE_SYSTEM",
+    0x00000038 : "FILE_DEVICE_TERMSRV",
+    0x00000021 : "FILE_DEVICE_TRANSPORT",
+    0x00000022 : "FILE_DEVICE_UNKNOWN",
+    0x0000002c : "FILE_DEVICE_VDM",
+    0x00000023 : "FILE_DEVICE_VIDEO",
+    0x00000024 : "FILE_DEVICE_VIRTUAL_DISK",
+    0x00000025 : "FILE_DEVICE_WAVE_IN",
+    0x00000026 : "FILE_DEVICE_WAVE_OUT",
+}
+
+vollog = logging.getLogger(__name__)
+
+class DeviceTree(interfaces.plugins.PluginInterface):
+    """Listing tree based on drivers and attached devices in a particular windows memory image."""
+
+    _required_framework_version = (2, 0, 1)
+    _version = (1, 0, 0)
+
+    @classmethod
+    def get_requirements(cls) -> List[interfaces.configuration.RequirementInterface]:
+        return [
+            requirements.ModuleRequirement(name = "kernel", description = "Windows kernel",
+                                           architectures = ["Intel32", "Intel64"]),
+            requirements.PluginRequirement(name = "driverscan", plugin = driverscan.DriverScan, version = (1, 0, 0)),
+        ]
+
+    def _generator(self) -> Iterator[Tuple]:
+        kernel = self.context.modules[self.config["kernel"]]
+
+        # Scan the Layer for drivers
+        for driver in driverscan.DriverScan.scan_drivers(self.context, kernel.layer_name, kernel.symbol_table_name):
+            try:
+                driver_name = driver.get_driver_name()
+
+                yield (0, (
+                    format_hints.Hex(driver.vol.offset),
+                    "DRV",
+                    driver_name,
+                    renderers.NotApplicableValue(),
+                    renderers.NotApplicableValue()
+                ))
+
+                # Scan to get the device information of driver.
+                for device in driver.get_devices():
+                    device_name = device.get_device_name()
+                    device_type = DEVICE_CODES.get(device.DeviceType, "UNKNOWN")
+
+                    yield (1, (
+                        format_hints.Hex(driver.vol.offset),
+                        "DEV",
+                        driver_name,
+                        device_name,
+                        device_type
+                    ))
+                    
+                    # Scan to get the attached devices information of device.
+                    for level, attached_device in enumerate(device.get_attached_devices(), start=2):
+                        device_name = attached_device.get_device_name()
+                        
+                        attached_device_name = "Unparsable Value" if isinstance(device_name, renderers.UnparsableValue) else device_name
```

Comment:
I had `extensions/__init__` return `UnparasableValue` if I could not get a specific data value for the drivers and devices. However, in this case, there was a phenomenon that the string formatting went into the object format.
```
<volatility3.framework.renderers.UnparsableValue object at 0x104efce80> - \Driver\USBHUB3
```
Therefore, I added the following logic by referring to the netscan plugin.
This requires verification whether the right logic. 😅

Reply:
I will solve this conversations with the under review!

Reply:
I see, that's fine, but it's the reason we don't use `BaseAbsentValues` until just before they're used.  This method should be returning None and then the check becomes for whether it's None or not (you can keep the "Unparsable value" if you want, rather than turning the entire thing into an unparsable value, but again, I'd suggest that each of these is a different item of data, and so should make up two columns rather than one in the output (and then the conversion wouldn't need to happen).

Reply:
That's right, as I worked on the feedback from the previous review, I thought it would be better to show the data in a separate column rather than merge the text. So if you are working on this and request a review again, please check it out! :)